controllers/dockermachine_controller.go

```diff
@@ -47,17 +68,257 @@ func (r *DockerMachineReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
+		return ctrl.Result{}, err
+	}
+
+	// TODO should the cluster be deleted?
+
+	return result, nil
+}
+
+func getRole(machine *capiv1alpha2.Machine) string {
+	// Figure out what kind of node we're making
+	labels := machine.GetLabels()
+	setValue, ok := labels["set"]
+	if !ok {
+		setValue = constants.WorkerNodeRoleValue
+	}
+	return setValue
+}
+
+func (r *DockerMachineReconciler) create(
+	ctx context.Context,
+	c *capiv1alpha2.Cluster,
+	machine *capiv1alpha2.Machine,
+	dockerMachine *infrastructurev1alpha2.DockerMachine,
+) (ctrl.Result, error) {
+	r.Log.Info("Creating a machine for cluster", "cluster-name", c.Name)
+	clusterExists, err := cluster.IsKnown(c.Name)
+	if err != nil {
+		r.Log.Error(err, "Error finding cluster-name", "cluster", c.Name)
+		return ctrl.Result{}, err
+	}
+	// If there's no cluster, requeue the request until there is one
+	if !clusterExists {
+		r.Log.Info("There is no cluster yet, waiting for a cluster before creating machines")
+		return ctrl.Result{
+			RequeueAfter: time.Second * 30,
+		}, nil
+	}
+	controlPlanes, err := actions.ListControlPlanes(c.Name)
+	if err != nil {
+		r.Log.Error(err, "Error listing control planes")
+		return ctrl.Result{}, err
+	}
+	r.Log.Info("Is there a cluster?", "cluster-exists", clusterExists)
+	setValue := getRole(machine)
+	r.Log.Info("This node has a role", "role", setValue)
+	if setValue == clusterAPIControlPlaneSetLabel {
+		// joining controlplane
+		if len(controlPlanes) > 0 {
+			r.Log.Info("Adding a control plane")
+			controlPlaneNode, err := actions.AddControlPlane(c.Name, machine.GetName(), *machine.Spec.Version)
+			if err != nil {
+				r.Log.Error(err, "Error adding control plane")
+				return ctrl.Result{}, err
+			}
+			providerID := actions.ProviderID(controlPlaneNode.Name())
+			dockerMachine.Spec.ProviderID = &providerID
+			dockerMachine.Status.Ready = true
+			return ctrl.Result{}, nil
+		}
+
+		// create a control plane
+		r.Log.Info("Creating a brand new cluster")
+		elb, err := getExternalLoadBalancerNode(c.Name, r.Log)
+		if err != nil {
+			r.Log.Error(err, "Error getting external load balancer node")
+			return ctrl.Result{}, err
+		}
+		lbipv4, _, err := elb.IP()
+		if err != nil {
+			r.Log.Error(err, "Error getting node IP address")
+			return ctrl.Result{}, err
+		}
+		controlPlaneNode, err := actions.CreateControlPlane(c.Name, machine.GetName(), lbipv4, *machine.Spec.Version, nil)
+		if err != nil {
+			r.Log.Error(err, "Error creating control plane")
+			return ctrl.Result{}, err
+		}
+		// set the machine's providerID
+		providerID := actions.ProviderID(controlPlaneNode.Name())
+		dockerMachine.Spec.ProviderID = &providerID
+		dockerMachine.Status.Ready = true
+
+		s, err := kubeconfigToSecret(c.Name, c.Namespace)
+		if err != nil {
+			r.Log.Error(err, "Error converting kubeconfig to a secret")
+			return ctrl.Result{}, err
+		}
+		// Save the secret to the management cluster
+		if err := r.Client.Create(ctx, s); err != nil {
```

Comment:
This is going to fail on an rbac issue, take a look here and add a similar line for the v1.Secrets with the create verb

https://github.com/kubernetes-sigs/cluster-api-provider-docker/pull/148/files#diff-5a0ef128e5d699a37a70d38f9a51d847R61-R62

Reply:
this change will require a regen of the manifests